src/lib.rs

```diff
@@ -91,25 +99,77 @@ fn get_output_mode(args: &Args) -> OutputMode {
     }
 }
 
+struct MaybeInitializedCaptureIndex(AtomicU32);
+
+impl MaybeInitializedCaptureIndex {
+    fn mark_failed(&self) {
+        self.0.store(u32::MAX - 1, Ordering::Relaxed);
+    }
+
+    pub fn get(&self) -> Result<Option<u32>, ()> {
+        let loaded = self.0.load(Ordering::Relaxed);
+        match loaded {
+            loaded if loaded == u32::MAX => Ok(None),
+            loaded if loaded == u32::MAX - 1 => Err(()),
+            loaded => Ok(Some(loaded)),
+        }
+    }
+
+    pub fn get_or_initialize(&self, query: &Query, capture_name: Option<&str>) -> Result<u32, ()> {
+        if let Some(already_initialized) = self.get()? {
+            return Ok(already_initialized);
+        }
+        let capture_index = match capture_name {
+            None => 0,
+            Some(capture_name) => {
+                let capture_index = query.capture_index_for_name(capture_name);
+                if capture_index.is_none() {
+                    self.mark_failed();
+                    fail(&format!("invalid capture name '{}'", capture_name));
```

Comment:
So in theory this would be the one time that this error message gets printed (other threads then calling `.get_or_initialize()` would early-return an `Err` above)

And in practice (eg I added a test for this) I've only been seeing it get printed once

However I think it's actually "race"-y because I think another thread could enter this block before this one has had a chance to `.mark_failed()` here

Ooh actually I bet I could use a "fetch-swap" or whatever in `.mark_failed()` so basically ensure that after the call to `.mark_failed()` it's definitely stored the failure value but then return a value from that call to `.mark_failed()` that indicates whether we actually "won" the race to insert that failure value. And then only print the failure message here if we won

Reply:
Ok I updated this to use `.compare_exchange()` to avoid such a race. And in fact "validated" it by artificially introducing a long sleep, seeing the race happen (ie multiple printings of the failure message), and then no longer seeing it happen after adding the `.compare_exchange()`